### PR TITLE
Data Preparation

### PR DESCRIPTION
**Pull request by pasztork**
I have implemented a module for preparing data for the detection task. Additionally, I have renamed Gergő's module, which was originally responsible for data preparation for the segmentation task, to better reflect its purpose.

Furthermore, I have developed a PyTorch Lightning `DataModule` that supports both detection and segmentation datasets.

### CODE REVIEW

File: src/data_preparation.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+import sys
+from typing import Dict, Tuple
+
+import cv2
+import numpy as np
+import torchvision.transforms.functional as TF
+from PIL import Image
+
+
+def zoom_image(image: Image.Image, zoom_factor: float = 1.5) -> Image.Image:
```
Comment (tmb2k01) [inline src/data_preparation.py]: I'm not sure why are we using these functions, when Pytorch Lightning has built-in augmentations, and it's accelerated with GPU.

File: src/data_preparation.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    transformations["clahe_zoom"] = TF.resize(img_zoom, target_size)
+
+    # Flip and zoom the image
+    img_flip_zoom = zoom_image(img_flip)
+    transformations["clahe_flip_zoom"] = TF.resize(img_flip_zoom, target_size)
+
+    return transformations
+
+
+def prepare_data(input_directory: str, output_directory: str, target_size: int) -> None:
```
Comment (tmb2k01) [inline src/data_preparation.py]: I'm missing the dataloader class, how can we load the images and give them directly to the model afterwards?

File: src/data_preparation.py
Comment (tmb2k01) [inline src/data_preparation.py:1]: The idea is good, but we should use Pytorch Lightning without these additional steps. This will generate a lot of code, and most of it will be thrown away.

Comment (pasztork) [reply src/data_preparation.py:1]: You're totally right. Will fix it soon

**Review by fustosgergo — APPROVED** [review]
Looks perfect 💯

File: src/data_module.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        super().__init__()
+
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+
+        self.train_dataset = CbisDdsmDataset(
+            root_dir=os.path.join(root_dir, "train"),
+            transform=transforms.Compose(
+                [
+                    RandomFlip(0.5, 0.5),
```
Comment (tmb2k01) [inline src/data_module.py:164]: Why dont we use from torchvision import transforms?

Comment (pasztork) [reply src/data_module.py:164]: Since those don't operate on the images the way we need it. We need to transform both the original images and the masks. If we rotated the input image, we should also have a rotated mask for evaluation, at least that is what I think. We do actually use Pytorch Transforms inside the custom transforms.

File: src/data_module.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    def __call__(self, sample):
+        image, masks = sample["image"], sample["masks"]
+
+        if np.random.random() < self.ph:
+            image = TF.hflip(image)
+            masks = [TF.hflip(m) for m in masks]
+
+        if np.random.random() < self.pv:
+            image = TF.vflip(image)
+            masks = [TF.hflip(m) for m in masks]
```
Comment (pasztork) [inline src/data_module.py]: This should be `TF.vflip`. Will fix.

File: src/data_module.py (hunk 3 above)
Comment (pasztork) [reply src/data_module.py:164]: # Addressing Concerns

## Using Custom Transform Classes

Quoting the official [documentation](https://pytorch.org/vision/0.18/datasets.html):

> All the datasets have almost similar API. They all have two common arguments: `transform` and `target_transform` to transform the input and target respectively.

In essence, this means that transformations are not automatically applied to the expected output (i.e., masks). We could use `target_transform` to transform the masks, but this would make it difficult to apply consistent random transformations to both the mammographic images and the masks. For instance, if a vertical flip is applied to the mammographic image but not to the masks (due to separate random values being generated for each), the resulting sample would be incorrect.

## Returning a `dict` Instead of a `tuple`

The [Pytorch Lightning documentation](https://lightning.ai/docs/pytorch/LTS/common/lightning_module.html) states:

> Under the hood, Lightning does the following (pseudocode):
```python
# put model in train mode and enable gradient calculation
model.train()
torch.set_grad_enabled(True)

outs = []
for batch_idx, batch in enumerate(train_dataloader):
    loss = training_step(batch, batch_idx)
    outs.append(loss.detach())

    # clear gradients
    optimizer.zero_grad()

    # backward
    loss.backward()

    # update parameters
```

This implies that we can return any data structure from the `__getitem__` method, as long as it is unpacked correctly in the corresponding method within the `LightningModule`. Therefore, returning a dictionary instead of a tuple should work without any issues.

However, some built-in Lightning features (like automatic logging, certain callbacks, or plugins) might assume a tuple-like `(input, target)` format. We may need to handle any such cases, but I'm unsure whether we'll use any such features. If you wouldn't like to have to deal with such issues, I'm more than happy to change up the current implementation. Let me know what you think 😃

Comment (tmb2k01) [reply src/data_module.py:164]: In my opinion, it's a more reasonable decision to use tuples because, if we change our minds in the future, fixing this issue would require much more work. Nice job btw! 😄

Comment (pasztork) [reply src/data_module.py:164]: Alright, I'll go ahead and update the module 🔨

Comment (pasztork) [reply src/data_module.py:164]: As promised, I’ve updated the data module and added a mock model so you can see how everything will work together. Does this look like what you had in mind for the data module?

**Review by tmb2k01 — APPROVED** [review]
